src/typing.py

```diff
@@ -1916,7 +1916,21 @@ def __new__(cls, typename, bases, ns):
             raise TypeError("Class syntax for NamedTuple is only supported"
                             " in Python 3.6+")
         types = ns.get('__annotations__', {})
-        return _make_nmtuple(typename, types.items())
+        nm_tpl = _make_nmtuple(typename, types.items())
+        saw_default = False
+        defaults = []
+        defaults_dict = {}
+        for field_name in types:
+            if field_name in ns:
+                saw_default = True
+                default_value = ns[field_name]
+                defaults.append(default_value)
+                defaults_dict[field_name] = default_value
+            elif saw_default:
+                raise TypeError('Non-default namedtuple field cannot follow default field')
```

Comment:
It could be sometimes helpful to mention which fields are conflicting, for example ``f'Non-default namedtuple field {field_name} cannot follow default fields {list(defaults_dict.keys())}'``.